### ISSUE TITLE
Uncaught ReferenceError: o10n is not defined

### ISSUE THREAD

**Issue by ahmed-sigmalux** [o10n-x/wordpress-css-optimization#29]
I'm seeing this error on all pages when logged in. I believe it's coming from:

`<script>o10n.constructor.prototype.extract=function(t){(function(d,c,s){s=d.createElement('script');s.async=true;s.onload=c;s.src="https:\/\/www.mydomain.com\/extensions\/wordpress-css-optimization\/public\/js\/critical-css-widget.min.js";d.head.appendChild(s);})(document,function(){o10n.extract(t);});}</script>`

**Comment by optimalisatie** [comment]
Thank you for reporting the issue!

The error indicates that you are using the CSS optimization plugin with javascript client related settings (e.g. async loading) disabled. This results in no requirement for the `o10n` javascript client which will then not be inserted into the HTML document.

As a admin however, the menu will provide a link to `Export Critical CSS` that extends the javascript client.

We will fix this issue in the future. It is a minor issue. 

If you want to use the critical CSS extraction while the CSS optimization plugin is inactive you could manually copy/paste the following code into the browser console:

```javascript
(function(d,c,s) {s=d.createElement('script');s.async=true;s.onload=c;s.src='https://cdn.rawgit.com/o10n-x/critical-css-widget/master/critical-css-widget.min.js';d.head.appendChild(s);})(document,function() {
   // critical css file download
   o10n.extract();
});
```

More information about the widget is available on https://github.com/o10n-x/critical-css-widget


### Custom 'modular' Google Closure Compiler

The javascript client of the plugin is pretty advanced to provide the maximum compression possible. The client of all plugins, including javascript and font optimization, can be stitched together as raw code without a module loader, while internally they reuse their code. This is made possible by Google Closure Compiler module compilation mode which appears to be a undocumented but high value feature. 

The plugin creates the smallest javascript client possible and doesn't waste a single byte for any given configuration. On top of that, it uses a compressed JSON format that is attached as `data-*` attribute which will enable to make the client script compliant with a `Content-Security-Policy`.

We created a custom grunt tool to enable modular compilation.

<details/>
  <summary>Custom Grunt.js module for Google Closure Compiler modular compilation</summary>

```javascript
module.exports = function(grunt) {
    'use strict';

    var exec = require('child_process').exec,
        fs = require('fs'),
        path = require('path');

    // ==========================================================================
    // TASKS
    // ==========================================================================

    grunt.registerMultiTask('o10n-closure-compiler', 'o10n Closure Compiler.', function() {

        var closurePath = '',
            reportFile = '',
            data = this.data,
            done = this.async();

        // Check for closure path.
        if (data.closurePath) {
            closurePath = data.closurePath;
        } else if (process.env.CLOSURE_PATH) {
            closurePath = process.env.CLOSURE_PATH;
        } else {
            grunt.log.error('' +
                '/!\\'.red +
                ' Set an environment variable called ' +
                'CLOSURE_PATH'.red + ' or the build parameter' + 'closurePath'.red +
                ' and\nmake it point to your root install of Closure Compiler.' +
                '\n');
            return false;
        }

        var command = 'java -jar "' + closurePath + '/build/compiler.jar"';

        data.cwd = data.cwd || './';

        if (data.externs) {
            data.externs = grunt.file.expand(data.externs);
            command += ' --externs ' + data.externs.join(' --externs ');

            if (!data.externs.length) {
                delete data.externs;
            }
        }

        if (data.options.externs) {
            data.options.externs = grunt.file.expand(data.options.externs);

            if (!data.options.externs.length) {
                delete data.options.externs;
            }
        }

        for (var directive in data.options) {
            if (Array.isArray(data.options[directive])) {
                command += ' --' + directive + ' "' + data.options[directive].join('" --' + directive + ' "') + '"';
            } else if (data.options[directive] === undefined || data.options[directive] === null) {
                command += ' --' + directive;
            } else {
                command += ' --' + directive + ' "' + String(data.options[directive]) + '"';
            }
        }

        for (var module in data.modules) {

            //command += '" --define="DEBUG=false"';
            data.modules[module].src.forEach(function(src) {
                command += ' --js "' + src + '"';
            });
            command += ' --module "' + module + ':' + data.modules[module].src.length + ':';
            if (data.modules[module].dep) {
                command += data.modules[module].dep.join(',');
                command += ':';
            }
            command += '"';
        }

        // save to dist/js/
        command += ' --module_output_path_prefix ' + ((data.debug) ? 'dist-debug/js/' : 'dist/js/');

        // because closure compiler does not create dirs.
        //grunt.file.write(data.jsOutputFile, '');

        // Minify WebGraph class.
        exec(command, {
            maxBuffer: data.maxBuffer * 1024,
            cwd: data.cwd
        }, function(err, stdout, stderr) {
            if (err) {
                grunt.warn(err);
                done(false);
            }

            if (stdout) {
                grunt.log.writeln(stdout);
            }

            // OK
            done();

        });

    });


};
```
</details>

**Comment by optimalisatie** [comment]
Hi @ahmed-sigmalux,

Thanks again for your feedback!

The issue has been resolved in `v0.0.88`.

https://github.com/o10n-x/wordpress-css-optimization/releases/tag/0.0.88
